validator.go

```diff
@@ -62,9 +62,9 @@ type validator struct {
 	expectedSub string
 }
 
-// newValidator can be used to create a stand-alone validator with the supplied
+// NewValidator can be used to create a stand-alone validator with the supplied
 // options. This validator can then be used to validate already parsed claims.
-func newValidator(opts ...ParserOption) *validator {
+func NewValidator(opts ...ParserOption) *validator {
```

Comment:
I think the intent was to keep this unexported because it's a massive foot gun.

Should we add a warning comment directing folks to the "typical" usage?

Reply:
Just to be on the same page. What are your main concerns with it? I thought it is pretty much designed in a way that you cannot mess up default validation, but maybe I am too optimistic? ;)

Reply:
We were funnelling users towards the Parser, which did all the right things. Something about exposing a raw validator without verification seems "wrong". But I don't think that's a good argument for _not_ doing this.

TL;DR - ignore this point, and instead, let's focus on refactoring this to avoid retuning an unexported identifier `*validator`

Reply:
> We were funnelling users towards the Parser, which did all the right things. Something about exposing a raw validator without verification seems "wrong". But I don't think that's a good argument for _not_ doing this.

Ah, got it!

> 
> TL;DR - ignore this point, and instead, let's focus on refactoring this to avoid retuning an unexported identifier `*validator`

Yes, definitely. So we are either stuck exporting `*Validator` (without any public struct elements) or a new interface. Since the parser is also already a exported struct and not an interface, I would lean towards a struct as well for some consistency.

Reply:
Agreed, I was leaning towards a struct as well.

Reply:
I wonder if there is an alternative to directly exporting this somewhat "problematic" struct directly in the `jwt` package. Since its main use case are probably (unit) tests, I wonder if we could introduce a `jwttest` package, similar to the *test package in the stdlib, such as https://pkg.go.dev/net/http/httptest. We could then export the Validator (or maybe a TestValidator) in the `jwttest` package. To do so we probably would need to move the `validator` to an `internal` package.

Not sure if this is cleaner in the end but it would probably irritate people less than having the `Validator` and `Parser` struct directly in `jwt`. What do you think?

Reply:
Indeed, this is a nice pattern (good idea). Let's keep this branch open, but open another to see how that would behave/feel.

I think there will always be a cohort of folks who'll want direct access to the Validator, but ideally, we gather a bit more feedback before committing to it.

Reply:
> I wonder if there is an alternative to directly exporting this somewhat "problematic" struct directly in the `jwt` package. Since its main use case are probably (unit) tests, I wonder if we could introduce a `jwttest` package, similar to the *test package in the stdlib, such as https://pkg.go.dev/net/http/httptest. We could then export the Validator (or maybe a TestValidator) in the `jwttest` package. To do so we probably would need to move the `Validator` to an `internal` package, so it can be accessed by us internally as well as by the `jwtttest` package.
> 
> Not sure if this is cleaner in the end but it would probably irritate people less than having the `Validator` and `Parser` struct directly in `jwt`. What do you think?

It's worth considering this at least, though I'm not sure `jwttest` is the best name for it.

Perhaps `jwtvalidator` or `jwtclaims`? (I was about to suggest `jwtonlyclaims`, but just thinking that name in my head made me kinda lol due to similarity to a popular website.)

Reply:
I did a quick prototype on this, but it seems that it is impossible because of the way our types are currently structured. If we move the validator to an extra package, we still need to access types such as `Claims` (which resides in `jwt` from the `validator` package. In turn, we need to access `validator.Validator` then from the `Parser`, which leads to a cyclic dependency.

The only way to avoid that would probably to move all of the common types to a common (internal?) package and then use a lot of type aliases, but I am not sure if that is really a good solution for this problem.

In that case I would rather swallow the pill and let it reside in `jwt`.

Reply:
> The only way to avoid that would probably to move all of the common types to a common (internal?) package and then use a lot of type aliases, but I am not sure if that is really a good solution for this problem.

That's what I was originally thinking, but, likely not worth it given the code complexity. Making the project harder to maintain isn't a good tradeoff.